Modify to identify only the changed lines from a given patch

kasunsiyambalapitiya opened the pull request:
Previous program considered a line range containing the modified lines from the given patch, for finding the authors, author commits and approved reviewers. As asked in the second code review the code is changed to detect exact lines been modified from the given patch and find authors, author commits and approved reviewers of those lines. This PR also contains all the modifications asked in PR1 [https://github.com/wso2-incubator/code-quality-matrices/pull/1](url)

Note:
At present the program works well for all WSO2 patches except the patches containing **octopus merge commits** ( commits having more than 2 parents) which may rarely available in WSO2 repositories

kasunsiyambalapitiya committed:
Correct the year in copyrights

kasunsiyambalapitiya committed:
added .idea direcotory to gitignore file

kasunsiyambalapitiya committed:
add entries for gitignore

kasunsiyambalapitiya committed:
removed unwanted files from the repository

kasunsiyambalapitiya committed:
Add log messages

kasunsiyambalapitiya committed:
Add log messages

Add log messages for all classes

kasunsiyambalapitiya committed:
Merge branch used for adding logs to the classes

kasunsiyambalapitiya committed:
Remove the CodeQualityMatrices.iml from remote

kasunsiyambalapitiya committed:
Get the patch details for a file made by the given commit

kasunsiyambalapitiya committed:
Complete upto getting authors of the buggly lines from SDK

kasunsiyambalapitiya committed:
Added docs

kasunsiyambalapitiya committed:
Pmt.java convert to use java8

kasunsiyambalapitiya committed:
convert BlameCommit.java to java8

kasunsiyambalapitiya committed:
create seperate class for calling graphql API

kasunsiyambalapitiya committed:
Change the Logger initiliation in classes as suggested

in the pull request https://github.com/kasunsiyambalapitiya/CodeQualityMatricesProject/pull/1

format the code

kasunsiyambalapitiya committed:
fix the copyright header

kasunsiyambalapitiya committed:
add CodeQualityMatricesException class

kasunsiyambalapitiya committed:
Add exceptions for classes

kasunsiyambalapitiya committed:
Remove invalid user input handling in RestApiCaller.java

kasunsiyambalapitiya committed:
Rename log object and fixed possible null value passings

kasunsiyambalapitiya committed:
fix the changes asked by DilanUA in PR 2

kasunsiyambalapitiya committed:
Changed method names to have a good English flow

kasunsiyambalapitiya committed:
change the part where it skips taking blme of new files

kasunsiyambalapitiya committed:
replaced code with java 8 features

added streams for saveReviewersToList method

added streams for savePrNumberAndRepoName method

replaced for loop with for each loop in readTheReviewOutJSON method

kasunsiyambalapitiya committed:
Replaced variables with contants

add constants for Reviewers.java

kasunsiyambalapitiya committed:
Correct Exceptional Handling

kasunsiyambalapitiya committed:
Delete replaced codes from Java 8

kasunsiyambalapitiya committed:
fix formatting in java 8 stream API codes

kasunsiyambalapitiya committed:
Handle response codes

Handle the response code recieved from Graphql API

Correct the message thrown in Exception

kasunsiyambalapitiya committed:
fix getting graphql blame issue

kasunsiyambalapitiya committed:
use stream API in Reviewers.java

kasunsiyambalapitiya committed:
changes within the code review

kasunsiyambalapitiya committed:
Seperate PMT communication to a seperate class

kasunsiyambalapitiya committed:
Seperate Github communication to a seperate class

kasunsiyambalapitiya committed:
Add a utility class

kasunsiyambalapitiya committed:
Use Gson for PMT response processing

kasunsiyambalapitiya committed:
Use Gson for processing Github Search commit API response

kasunsiyambalapitiya committed:
Add model classes

kasunsiyambalapitiya committed:
use Gson till getting till getting author commits

kasunsiyambalapitiya committed:
Change name of Graphql Pojo classes

kasunsiyambalapitiya committed:
Add pojo classes for github Issue API

kasunsiyambalapitiya committed:
Use Gson for processing review API output

kasunsiyambalapitiya committed:
Add java docs

kasunsiyambalapitiya committed:
Delete unused classes

kasunsiyambalapitiya committed:
Fix bugs found in running Idea Analyse

kasunsiyambalapitiya committed:
Fix groupId in pom.xml

kasunsiyambalapitiya committed:
Fix errors received by adding wso2 parent pom

kasunsiyambalapitiya committed:
Remove .classpath and .project created from Eclipse IDE

kasunsiyambalapitiya committed:
Moved .gitignore a level up in directory

kasunsiyambalapitiya committed:
Change the repo name

kasunsiyambalapitiya committed:
fix changes asked by Nisala in PR 1

kasunsiyambalapitiya committed:
fix changes asked in PR 1

kasunsiyambalapitiya committed:
Merge branch 'annotatePojoClass' into changesFromCodeReview2

kasunsiyambalapitiya committed:
Rename the package

kasunsiyambalapitiya committed:
Use a Configuration file

kasunsiyambalapitiya committed:
Add test cases

kasunsiyambalapitiya committed:
fix for considering  only deleted lines

kasunsiyambalapitiya committed:
Delete unnecassary lines

kasunsiyambalapitiya committed:
Check for octopuss commits

kasunsiyambalapitiya committed:
Add testcases for new approach

Shan1024 commented on org.wso2.codequalitymatrices/src/main/java/com/wso2/code/quality/metrics/ReviewAnalyser.java:
```diff
@@ -0,0 +1,174 @@
+                        try {
+                            String jsonText = githubApiCaller.callReviewApi(repositoryName, prNumber, githubToken);
+                            if (jsonText != null) {
+//                                Type listType = new TypeToken<List<ReviewApiResponse>>() {
```
Remove commented codes.

kasunsiyambalapitiya replied:
Fixed in 698c71ed0e32960930e00a53d3272028ab2d399a

Shan1024 commented on org.wso2.codequalitymatrices/src/main/java/com/wso2/code/quality/metrics/model/Author.java:
```diff
@@ -0,0 +1,41 @@
+public class Author {
+
+    @SerializedName("date")
+
```
Unnecessary newline.

kasunsiyambalapitiya replied:
Fixed in 698c71ed0e32960930e00a53d3272028ab2d399a

Shan1024 commented on org.wso2.codequalitymatrices/src/main/java/com/wso2/code/quality/metrics/model/CommitHistoryApiResponse.java:
```diff
@@ -0,0 +1,51 @@
+    private String commitHash;
+
+    @SerializedName("commit")
+
```
Unnecessary newline.

kasunsiyambalapitiya replied:
Fixed in 698c71ed0e32960930e00a53d3272028ab2d399a

Shan1024 commented on org.wso2.codequalitymatrices/src/test/java/com/wso2/code/quality/metrics/GithubResponsesTest.java:
```diff
@@ -0,0 +1,114 @@
+    @Test
+    public void testSaveRepoNames() throws CodeQualityMetricsException {
+        Map<String, List<String>> commitHashWithRepoNames = new HashMap<>();
+        commitHashWithRepoNames.put("eaa45529cbabc5f30a2ffaa4781821ad0a5223ab",
```
These hashes seems to be used in multiple places. So might be able to use a constant.

chamilad commented on org.wso2.codequalitymatrices/src/main/java/com/wso2/code/quality/metrics/ApiUtility.java:
```diff
@@ -0,0 +1,171 @@
+/**
+ * This is a utility class for calling APIs.
+ *
+ * @since 1.0.0
```
This doc line is not used AFAIR. Better to check for consistency with other repositories.

kasunsiyambalapitiya replied:
I checked with few WSO2 repos, some contain @since tag some doesn't, since this is a class level doc comment I think it is better to have it, What is your opinion?

Shan1024 replied:
@kasunsiyambalapitiya : I think what chamila meant was don't use "This" word in the doc section. We all know that the doc is related to "This". So we can just say "Utility class for calling APIs".

chamilad commented on org.wso2.codequalitymatrices/src/main/java/com/wso2/code/quality/metrics/ApiUtility.java:
```diff
@@ -0,0 +1,171 @@
+        BufferedReader bufferedReader = null;
+        CloseableHttpClient httpClient;
+        CloseableHttpResponse httpResponse = null;
+        httpClient = HttpClients.createDefault();
```
Any reason not to instantiate at the same line where this is declared?

kasunsiyambalapitiya replied:
Fixed in 698c71ed0e32960930e00a53d3272028ab2d399a

chamilad commented on org.wso2.codequalitymatrices/src/main/java/com/wso2/code/quality/metrics/ApiUtility.java:
```diff
@@ -0,0 +1,171 @@
+        String jsonText;
+        try {
+            httpResponse = httpClient.execute(httpGet);
+            int responseCode = httpResponse.getStatusLine().getStatusCode();
```
Is it possible for `httpResponse.getStatusLine()` to return `null`? If so this might be a possible NPE producer.

kasunsiyambalapitiya replied:
Handled NPE in 698c71ed0e32960930e00a53d3272028ab2d399a

chamilad commented on org.wso2.codequalitymatrices/src/main/java/com/wso2/code/quality/metrics/ApiUtility.java:
```diff
@@ -0,0 +1,171 @@
+            int responseCode = httpResponse.getStatusLine().getStatusCode();
+            if (responseCode == 200) {
+                //success
+                bufferedReader = new BufferedReader(new InputStreamReader(httpResponse.getEntity().getContent(),
```
Same with `httpResponse.getEntity()`. When chaining calls, better to verify for possible NPEs.

kasunsiyambalapitiya replied:
Fixed in 698c71ed0e32960930e00a53d3272028ab2d399a

chamilad commented on org.wso2.codequalitymatrices/src/main/java/com/wso2/code/quality/metrics/Application.java:
```diff
@@ -0,0 +1,47 @@
+        if (args.length == 3) {
+            String pmtToken = args[0];
+            String patchId = args[1];
+            String gitHubToken = args[2];
```
IMHO, it might be better to offload tokens to a config file and only accept the patch ID as an argument, as tokens tend to be reused again and would be cumbersome to enter in the command line.

kasunsiyambalapitiya replied:
Fixed in 698c71ed0e32960930e00a53d3272028ab2d399a

chamilad commented on org.wso2.codequalitymatrices/src/main/java/com/wso2/code/quality/metrics/GithubApiCaller.java:
```diff
@@ -0,0 +1,213 @@
+ *
+ * @since 1.0.0
+ */
+public class GithubApiCaller {
```
This looks like a util class and could be final and methods could be static as there doesn't seem to be any state to be preserved between calls.

kasunsiyambalapitiya replied:
Converted to a Util class in 698c71ed0e32960930e00a53d3272028ab2d399a

chamilad commented on org.wso2.codequalitymatrices/src/main/java/com/wso2/code/quality/metrics/PmtApiCaller.java:
```diff
@@ -0,0 +1,68 @@
+ * @since 1.0.0
+ */
+public class PmtApiCaller {
+    public PmtApiCaller() {
```
Why is this constructor declared? Leftover from any refactoring?

kasunsiyambalapitiya replied:
:) Removed in 698c71ed0e32960930e00a53d3272028ab2d399a

chamilad commented on org.wso2.codequalitymatrices/src/main/java/com/wso2/code/quality/metrics/ReviewAnalyser.java:
```diff
@@ -0,0 +1,174 @@
+     * @return a map of pull requests againt their repository name
+     * @throws CodeQualityMetricsException results
+     */
+    Map<String, Set<Integer>> savePrNumberAndRepoName(String jsonText) throws CodeQualityMetricsException {
```
Should this be default access? Can it be private?

kasunsiyambalapitiya replied:
Since this method is called from a test case private access cannot be given

Shan1024 replied:
@kasunsiyambalapitiya : You can give private access even it is used in a test case. AFAIK, you can use java reflection to temporary change the access modifier before calling this method from the test case.

chamilad commented on org.wso2.codequalitymatrices/src/main/java/com/wso2/code/quality/metrics/model/Constants.java:
```diff
@@ -0,0 +1,34 @@
+ * under the License.
+ */
+
+package com.wso2.code.quality.metrics.model;
```
If this class contains all the constants used in the project, then it might belong at the parent level of the package hierarchy, or under a separate `util` package.

chamilad commented on org.wso2.codequalitymatrices/src/main/java/com/wso2/code/quality/metrics/model/Graphql.java:
```diff
@@ -0,0 +1,57 @@
+    private String graphqlInputWithoutHistory;
+    private static final long serialVersionUID = 42L;
+
+    public Graphql() {
```
Unnecessary declaration?

kasunsiyambalapitiya replied:
Without it maven build fails giving this error `com.wso2.code.quality.metrics.model.Graphql is Serializable; consider declaring a serialVersionUID [com.wso2.code.quality.metrics.model.Graphql] At Graphql.java:[lines 32-44] SE_NO_SERIALVERSIONID`. I fixed it as mentioned in the answer by "Jon Skeet" in [http://stackoverflow.com/questions/285793/what-is-a-serialversionuid-and-why-should-i-use-it](url) or else we need to avoid implementing `Serializable` interface. What is the best way to handle it?

chamilad commented on org.wso2.codequalitymatrices/src/main/java/com/wso2/code/quality/metrics/model/GraphqlCommit.java:
```diff
@@ -0,0 +1,59 @@
+public class GraphqlCommit {
+    @SerializedName("author")
+    private GraphqlAuthor author;
+    @SerializedName("history")
```
Better to keep a new line between properties to improve readability.

kasunsiyambalapitiya replied:
Fixed in 698c71ed0e32960930e00a53d3272028ab2d399a

chamilad commented on org.wso2.codequalitymatrices/src/main/resources/log4j.properties:
```diff
@@ -27,10 +29,10 @@ log4j.appender.APPENDER1.MaxBackupIndex=5
+log4j.appender.CONSOLE.layout.conversionPattern=%d{yyyy-MM-dd}-5p--10c:%m%n
 
 
 
```
Unnecessary new lines?

kasunsiyambalapitiya replied:
Fixed in 698c71ed0e32960930e00a53d3272028ab2d399a

chamilad commented on org.wso2.codequalitymatrices/src/test/java/com/wso2/code/quality/metrics/Token.java:
```diff
@@ -0,0 +1,42 @@
+ * @since 1.0.0
+ */
+public class Token {
+    final String pmtToken = "tQU5vxzrGeBpLMQuwOsJW_fyYLYa";
```
Can't this be extracted out to the properties file? Looks like a hard coded config value.

kasunsiyambalapitiya replied:
Fixed in 698c71ed0e32960930e00a53d3272028ab2d399a

chamilad reviewed (changes requested):
Were you able to run any code coverage runs for the project? Do the unit tests cover important logic?

thusithathilina commented on org.wso2.codequalitymatrices/src/main/java/com/wso2/code/quality/metrics/ApiUtility.java:
```diff
@@ -0,0 +1,171 @@
+            int responseCode = httpResponse.getStatusLine().getStatusCode();
+            if (responseCode == 200) {
+                //success
+                bufferedReader = new BufferedReader(new InputStreamReader(httpResponse.getEntity().getContent(),
```
Can we use try with resource style for readers. Then you don't need to close them explicitly

kasunsiyambalapitiya replied:
Fixed in 698c71ed0e32960930e00a53d3272028ab2d399a

thusithathilina commented on org.wso2.codequalitymatrices/src/main/java/com/wso2/code/quality/metrics/ApiUtility.java:
```diff
@@ -0,0 +1,171 @@
+     * @throws CodeQualityMetricsException results
+     */
+    public static String callGraphQlApi(HttpPost httpPost) throws CodeQualityMetricsException {
+        BufferedReader bufferedReader = null;
```
If possible use java7 try with resource style

kasunsiyambalapitiya replied:
Fixed in 698c71ed0e32960930e00a53d3272028ab2d399a

thusithathilina commented on org.wso2.codequalitymatrices/src/main/java/com/wso2/code/quality/metrics/GithubApiCaller.java:
```diff
@@ -0,0 +1,213 @@
+        }
+        return ApiUtility.callGraphQlApi(httpPost);
+    }
+}
```
why add call to al the methods? can we think of something like search, getHistory etc..

thusithathilina commented on org.wso2.codequalitymatrices/src/main/java/com/wso2/code/quality/metrics/PmtApiCaller.java:
```diff
@@ -0,0 +1,68 @@
+        try {
+            Properties defaultProperties = new Properties();
+            ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+            InputStream inputStream = classLoader.getResourceAsStream("url.properties");
```
Need to close the InputStreams

kasunsiyambalapitiya replied:
Fixed in 698c71ed0e32960930e00a53d3272028ab2d399a

thusithathilina commented on org.wso2.codequalitymatrices/src/main/java/com/wso2/code/quality/metrics/ReviewAnalyser.java:
```diff
@@ -0,0 +1,174 @@
+            IssueApiResponse issueApiResponse = gson.fromJson(jsonText, IssueApiResponse.class);
+            issueApiResponse.getIssue().parallelStream()
+                    .filter(searchItem -> GITHUB_REVIEW_API_CLOSED_STATE.equals(searchItem.getStateOfThePr()))
+
```
remove blank line

kasunsiyambalapitiya replied:
Fixed in 698c71ed0e32960930e00a53d3272028ab2d399a

daneshk commented on org.wso2.codequalitymatrices/src/test/java/com/wso2/code/quality/metrics/ChangesFinderTest.java:
```diff
@@ -0,0 +1,97 @@
+            result = IOUtils.toString(classLoader.getResourceAsStream(path));
+        } catch (IOException e) {
+            logger.debug(e.getMessage(), e.getCause());
+        }
```
why exception is ignored here. better throw the exception. if we need to ignore, log it as an error.

kasunsiyambalapitiya replied:
Fixed in 698c71ed0e32960930e00a53d3272028ab2d399a

daneshk commented on org.wso2.codequalitymatrices/src/main/java/com/wso2/code/quality/metrics/model/SearchItem.java:
```diff
@@ -0,0 +1,59 @@
+ */
+public class SearchItem {
+    @SerializedName("state")
+    private String stateOfThePr;
```
better if we can change the variable name to prState

kasunsiyambalapitiya replied:
Fixed in 698c71ed0e32960930e00a53d3272028ab2d399a

kasunsiyambalapitiya committed:
Changes asked in PR

close resources with try with resources

Seperate tokens to a property file

Fix bug from history simplificatio

If two commits in the history of a file cancel each other out (the changes are reversed), then Git detect this and simplifies the git log output of the file by leaving out those two commits.

kasunsiyambalapitiya committed:
Add README.md file

Shan1024 commented on README.md:
```diff
@@ -0,0 +1,48 @@
+## Installation
+### Prerequisites
+* [Git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git) installed.
+* [Java](http://www.oracle.com/technetwork/java/javase/downloads/index-jsp-138363.html) installed
```
Missing . at the end.

Shan1024 commented:
@kasunsiyambalapitiya : Please remove the `org.wso2.codequalitymatrices` directory from the project root as well(it would be better to have **src, pom.xml, LICENSE and README** in the project root). Please refer [1] for more details about the directory layout of a maven project. 

[1] [Introduction to the Standard Directory Layout](https://maven.apache.org/guides/introduction/introduction-to-the-standard-directory-layout.html)

kasunsiyambalapitiya committed:
Use Java Reflections

kasunsiyambalapitiya committed:
Move Files outside
